torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -520,8 +527,22 @@ def free_unsharded_param(self) -> None:
         ):
             free_storage(tensor)
         if ca.compiled_autograd_enabled:
-            self.all_gather_outputs = []
-            self._unsharded_inner_tensors = []
+            # Under compile, `self.all_gather_outputs` and `self._unsharded_inner_tensors` are emptied
+            # right after every AllGather. So we need to free `self.unsharded_param` memory by
+            # looking into its own tensor storage instead.
+            unsharded_param_data = (
+                self.unsharded_param.data
```

Comment:
I think generally, calling `.data` is not specifically what we want. Why not call `.detach()`? Or actually, why do we even need to detach from the graph at all (is this needed for tracing)?

E.g., can we case on wrapper subclass before calling `.data`?

Reply:
sg - `self.unsharded_param` is a `nn.Parameter` which is a subclass on its own, so I think we might need `.data` to get its underlying tensor (could still be a subclass (e.g. DTensor) or plain tensor). I changed it so that we only do `.data` access for the subclass case